bindgen/callbacks.rs

```diff
@@ -62,7 +62,16 @@ pub trait ParseCallbacks: fmt::Debug {
     /// The first parameter represents the name and argument list (including the
     /// parentheses) of the function-like macro. The second parameter represents
     /// the expansion of the macro as a sequence of tokens.
-    fn func_macro(&self, _name: &str, _value: &[&[u8]]) {}
+    fn func_macro(&self, _name: &str, _args: &[&str], _value: &[&[u8]]) {}
+
+    /// Specify the type of a macro argument.
+    ///
+    /// This is needed if you want to generate a function instead of a macro.
+    /// If all argument types and the return type of a macro can be inferred,
+    /// a function will be generated instead of a macro.
+    fn func_macro_arg_type(&self, _name: &str, _arg: &str) -> Option<String> {
```

Comment:
Say you have a macro like

```c
#define times(x, y) (x * y)
```

This will normally generate

```rust
macro_rules! times {
  ($x:expr, $y:expr) => {{ $x * $y }};
}
```

If you specify the types for each argument with `func_macro_arg_type`, i.e. 

`func_macro_arg_type("times", "x") == Some("int")` 

and

`func_macro_arg_type("times", "y") == Some("int")`

it will generate a function instead, since all types can be resolved:

```rust
fn times(x: c_int, y: c_int) {
  x * y
}
```

Reply:
Hmmm interesting. What happens if you use both callbacks with the same macro?

Reply:
You mean both `func_macro` and `func_macro_arg_type`?

Nothing will happen. `func_macro` let's the user do whatever they want with the macro as input, it doesn't have any effect on what is generated. `func_macro` will probably not be needed as much since many macros which previously needed some custom generation are supported now.

Reply:
In practice you would only use `func_macro` for macros which cannot be parsed or cannot be generated by `cmacro`. Improving `cmacro` would then of course be the better alternative to using this function.

Reply:
Maybe we can just remove `func_macro` since `cmacro` supports function-like macros while `cexpr` didn't.